File: src/algorithms/pathfinding/flood-fill/flood-fill-bfs/sources/flood-fill-bfs.py
from collections import deque
from typing import List, Tuple, Dict


# Flood Fill BFS — classic paint bucket fill using breadth-first search
def flood_fill_bfs(
    grid: List[List[dict]],
    start: Tuple[int, int],
) -> Dict:
    row_count = len(grid)  # @step:initialize
    col_count = len(grid[0])  # @step:initialize
    filled_set = [[False] * col_count for _ in range(row_count)]  # @step:initialize
    filled = []  # @step:initialize
    # Seed the queue with the start cell
    queue = deque([(start[0], start[1])])  # @step:initialize,open-node
    filled_set[start[0]][start[1]] = True  # @step:open-node

    while queue:
        # Dequeue the front cell — BFS processes cells level by level
        current_row, current_col = queue.popleft()  # @step:close-node
        filled.append((current_row, current_col))  # @step:close-node

        # Explore 4-directional neighbors
        for delta_row, delta_col in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
            neighbor_row = current_row + delta_row
            neighbor_col = current_col + delta_col
            if neighbor_row < 0 or neighbor_row >= row_count:
                continue
            if neighbor_col < 0 or neighbor_col >= col_count:
                continue
            if grid[neighbor_row][neighbor_col]["type"] == "wall":
                continue
            if filled_set[neighbor_row][neighbor_col]:
                continue
            # Mark on enqueue to avoid duplicates
            filled_set[neighbor_row][neighbor_col] = True  # @step:open-node
            queue.append((neighbor_row, neighbor_col))  # @step:open-node

    return {"filled": filled, "count": len(filled)}  # @step:complete

File: src/algorithms/pathfinding/flood-fill/flood-fill-bfs/sources/FloodFillBfs.cpp
// Flood Fill BFS — classic paint bucket fill using breadth-first search
#include <queue>
#include <string>
#include <vector>

enum class CellType { Empty, Wall, Start, End };

struct GridCell {
  int row;
  int col;
  CellType cellType;
  std::string state;
};

struct FloodFillResult {
  std::vector<std::pair<int, int>> filled;
  int count;
};

FloodFillResult floodFillBfs(const std::vector<std::vector<GridCell>>& grid,
                              std::pair<int, int> start) {
  int rowCount = static_cast<int>(grid.size()); // @step:initialize
  int colCount = rowCount > 0 ? static_cast<int>(grid[0].size()) : 0; // @step:initialize
  std::vector<std::vector<bool>> filledSet(rowCount, std::vector<bool>(colCount, false)); // @step:initialize
  std::vector<std::pair<int, int>> filled; // @step:initialize
  // Seed the queue with the start cell
  std::queue<std::pair<int, int>> bfsQueue; // @step:initialize,open-node
  bfsQueue.push(start); // @step:initialize,open-node
  filledSet[start.first][start.second] = true; // @step:open-node

  const int deltaRows[] = {-1, 1, 0, 0};
  const int deltaCols[] = {0, 0, -1, 1};

  while (!bfsQueue.empty()) {
    // Dequeue the front cell — BFS processes cells level by level
    auto current = bfsQueue.front(); // @step:close-node
    bfsQueue.pop(); // @step:close-node
    int currentRow = current.first; // @step:close-node
    int currentCol = current.second; // @step:close-node
    filled.push_back({currentRow, currentCol}); // @step:close-node

    // Explore 4-directional neighbors (up, down, left, right)
    for (int dirIndex = 0; dirIndex < 4; dirIndex++) {
      int neighborRow = currentRow + deltaRows[dirIndex];
      int neighborCol = currentCol + deltaCols[dirIndex];
      if (neighborRow < 0 || neighborRow >= rowCount || neighborCol < 0 || neighborCol >= colCount)
        continue;
      if (grid[neighborRow][neighborCol].cellType == CellType::Wall) continue;
      if (filledSet[neighborRow][neighborCol]) continue;
      // Mark on enqueue to avoid duplicates
      filledSet[neighborRow][neighborCol] = true; // @step:open-node
      bfsQueue.push({neighborRow, neighborCol}); // @step:open-node
    }
  }
  int count = static_cast<int>(filled.size());
  return {filled, count}; // @step:complete
}
